sycl/source/detail/kernel_program_cache.hpp

```diff
@@ -203,11 +211,93 @@ class KernelProgramCache {
   ///
   /// This member function should only be used in unit tests.
   void reset() {
+    std::lock_guard<std::mutex> L1(MProgramCacheMutex);
+    std::lock_guard<std::mutex> L2(MKernelsPerProgramCacheMutex);
+    std::lock_guard<std::mutex> L3(MKernelFastCacheMutex);
     MCachedPrograms = ProgramCache{};
     MKernelsPerProgramCache = KernelCacheT{};
     MKernelFastCache = KernelFastCacheT{};
   }
 
+  /// Try to fetch entity (kernel or program) from cache. If there is no such
+  /// entity try to build it. Throw any exception build process may throw.
+  /// This method eliminates unwanted builds by employing atomic variable with
+  /// build state and waiting until the entity is built in another thread.
+  /// If the building thread has failed the awaiting thread will fail either.
+  /// Exception thrown by build procedure are rethrown.
+  ///
+  /// \tparam RetT type of entity to get
+  /// \tparam ExceptionT type of exception to throw on awaiting thread if the
+  ///         building thread fails build step.
+  /// \tparam KeyT key (in cache) to fetch built entity with
+  /// \tparam AcquireFT type of function which will acquire the locked version
+  /// of
+  ///         the cache. Accept reference to KernelProgramCache.
+  /// \tparam GetCacheFT type of function which will fetch proper cache from
+  ///         locked version. Accepts reference to locked version of cache.
+  /// \tparam BuildFT type of function which will build the entity if it is not
+  /// in
+  ///         cache. Accepts nothing. Return pointer to built entity.
+  ///
+  /// \return a pointer to cached build result, return value must not be
+  /// nullptr.
+  template <typename ExceptionT, typename GetCachedBuildFT, typename BuildFT>
+  auto getOrBuild(GetCachedBuildFT &&GetCachedBuild, BuildFT &&Build) {
+    using BuildState = KernelProgramCache::BuildState;
+    constexpr size_t MaxAttempts = 2;
+    for (size_t AttemptCounter = 0;; ++AttemptCounter) {
+      auto [BuildResult, InsertionTookPlace] = GetCachedBuild();
```

Comment:
`InsertionTookPlace` seems to be unused, but we talk about insertion in the comments. Are we missing a check?

Reply:
Also, using just `InsertionTookPlace` is not quite correct with how the code is structured now. In the case when the building thread fails with a non-SYCL exception, the other threads are notified to restart the build loop (via transition to `BS_Initial`). Note in this case the cache is not reset. This means that next time through the loop for all the other threads their `InsertionTookPlace` would be false, so we wouldn't be able to use that to get one of those threads to start building the `BuildResult`. (If the thread that failed initial remove the `BuildResult` from the map, then we could use `InseritonTookPlace` to get one thread to pick up the build.)